lib/datadog/di/probe_notifier_worker.rb

```diff
@@ -0,0 +1,244 @@
+            sleep 0.25
+            next
+          elsif queues_empty
+            break
+          else
+            sleep 0.25
+            next
+          end
+        end
+      end
+
+      private
+
+      attr_reader :transport
+      attr_reader :wake
+      attr_reader :thread
+
+      # This method should be called while @lock is held.
+      def io_in_progress?
+        @io_in_progress
+      end
+
+      attr_reader :last_sent
+
+      [
+        [:status, 'probe status'],
+        [:snapshot, 'snapshot'],
+      ].each do |(event_type, event_name)|
+        attr_reader "#{event_type}_queue"
+
+        # Adds a status or a snapshot to the queue to be sent to the agent
+        # at the next opportunity.
+        #
+        # If the queue is too large, the event will not be added.
+        #
+        # Signals the background thread to wake up (and do the sending)
+        # if it has been more than 1 second since the last send of the same
+        # event type.
+        define_method("add_#{event_type}") do |event|
+          @lock.synchronize do
+            queue = send("#{event_type}_queue")
+            # TODO determine a suitable limit via testing/benchmarking
+            if queue.length > 100
+              logger.warn("#{self.class.name}: dropping #{event_type} because queue is full")
+            else
+              queue << event
+            end
+          end
+
+          # Figure out whether to wake up the worker thread.
+          # If minimum send interval has elapsed since the last send,
+          # wake up immediately.
+          @lock.synchronize do
+            unless @wake_scheduled
+              @wake_scheduled = true
+              set_sleep_remaining
+              wake.signal
+            end
+          end
+        end
+
+        # Determine how much longer the worker thread should sleep
+        # so as not to send in less than MIN_SEND_INTERVAL since the last send.
+        # Important: this method must be called when @lock is held.
+        #
+        # Returns the time remaining to sleep.
+        def set_sleep_remaining
+          now = Core::Utils::Time.get_time
+          @sleep_remaining = if last_sent
+            [last_sent + MIN_SEND_INTERVAL - now, 0].max
+          else
+            0
+          end
+        end
+
+        public "add_#{event_type}"
+
+        # Sends pending probe statuses or snapshots.
+        #
+        # This method should ideally only be called when there are actually
+        # events to send, but it can be called when there is nothing to do.
+        # Currently we only have one wake-up signaling object and two
+        # types of events. Therefore on most wake-ups we expect to only
+        # send one type of events.
+        define_method("maybe_send_#{event_type}") do
+          batch = nil
+          @lock.synchronize do
+            batch = instance_variable_get("@#{event_type}_queue")
+            instance_variable_set("@#{event_type}_queue", [])
```

Comment:
Minor: It may be worth adding an `attr_accessor` and then use `send(...)` to access these variables.

Reply:
The reading side is done with an attribute, are you concerned that there maybe a spelling mistake here and the wrong variable will be written to?

Reply:
I was mostly thinking that `instance_variable_set` and `instance_variable_get` are very sharp weapons so having the attr_accessor seems a bit easier to avoid bugs.

On the other hand, it is true that we can misspell the creation of the attr_accessor as well, so it's not like there's no potential for bugs there either.

Reply:
Let me leave it this way for now but if the matter comes up again I can redo as attr_accessor.